File: Binary Tree/二叉树的遍历方式/102.二叉树的层序遍历.cpp
#include <vector>
#include <stack>
#include <queue>

using namespace std;
/*
 * @lc app=leetcode.cn id=102 lang=cpp
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> result;
        queue<TreeNode *> que;

        if (root != nullptr)
        {
            que.push(root);
        }

        while (!que.empty())
        {

            vector<int> cur;
            int size = que.size();

            // 使用size，不用que.size，因为que被遍历pop之后，size会不断变化
            for (int i = 0; i < size; i++)
            {
                TreeNode *node = que.front();
                que.pop();
                cur.push_back(node->val);

                if (node->left)
                {
                    que.push(node->left);
                }

                if (node->right)
                {
                    que.push(node->right);
                }
            }
            result.push_back(cur);
        }

        return result;
    }
};
// @lc code=end

File: Binary Tree/二叉树的遍历方式/102.二叉树的层序遍历.ts
/*
 * @lc app=leetcode.cn id=102 lang=typescript
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
  let currentQueue = [root]
  let nextQueue = []
  let result = []
  let child = []

  if (!root) {
    return []
  }

  while (currentQueue.length > 0) {
    let firstItem = currentQueue.shift()

    if (firstItem) {
      // 记录当前层级的数据
      child.push(firstItem.val)
      // 保留下一个层级的数据
      firstItem.left && nextQueue.push(firstItem.left)
      firstItem.right && nextQueue.push(firstItem.right)
    }

    // 遍历完当前层级
    if (currentQueue.length === 0) {
      // 当前层级的数据追加进二维数组
      result.push(child)
      child = []
      // 下一层级赋值给当前层级
      currentQueue = nextQueue
      // 清空下一层级
      nextQueue = []
    }
  }

  return result
};
// @lc code=end

